addons/overheating/functions/fnc_jamWeapon.sqf

```diff
@@ -9,74 +9,58 @@
+#include "script_component.hpp"
 
-EXPLODE_2_PVT(_this,_unit,_weapon);
+params ["_unit", "_weapon"];
+TRACE_2("params",_unit,_weapon);
 
 // don't jam a weapon with no rounds left
-if (_unit ammo _weapon == 0) exitWith {};
+private _ammo = _unit ammo _weapon;
+if (_ammo == 0) exitWith {};
 
-private ["_jammedWeapons"];
-_jammedWeapons = _unit getVariable [QGVAR(jammedWeapons), []];
+private _jammedWeapons = _unit getVariable [QGVAR(jammedWeapons), []];
 _jammedWeapons pushBack _weapon;
 
 _unit setVariable [QGVAR(jammedWeapons), _jammedWeapons];
 
 
-// this is to re-activate the 'DefaultAction', so you can jam a weapon while full auto shooting
-_fnc_stopCurrentBurst = {
-  EXPLODE_2_PVT(_this,_params,_pfhId);
-  EXPLODE_4_PVT(_params,_unit,_weapon,_ammo,_startFrame);
-
-  // Skip the first execution of the PFH
-  if (diag_frameno == _startFrame) exitWith {};
-
-  // Remove the PFH on the second execution
-  [_pfhId] call CBA_fnc_removePerFrameHandler;
-
-  _unit setAmmo [_weapon, _ammo];
-};
 
 // Stop current burst
-_ammo = _unit ammo _weapon;
 if (_ammo > 0) then {
     _unit setAmmo [_weapon, 0];
-    [_fnc_stopCurrentBurst, 0, [_unit, _weapon, _ammo, diag_frameno]] call CBA_fnc_addPerFrameHandler;
+    // this is to re-activate the 'DefaultAction', so you can jam a weapon while full auto shootin
+    [{
+        params ["_unit", "_weapon", "_ammo"];
+        _unit setAmmo [_weapon, _ammo];
+    }, [_unit, _weapon, _ammo]] call EFUNC(common,execNextFrame);
 };
 
 // only display the hint once, after you try to shoot an already jammed weapon
 GVAR(knowAboutJam) = false;
 
 if (_unit getVariable [QGVAR(JammingActionID), -1] == -1) then {
-  private ["_condition", "_statement", "_condition2", "_statement2", "_id"];
-
-  _condition = {
-    [_this select 1] call EFUNC(common,canUseWeapon)
-    && {currentMuzzle (_this select 1) in ((_this select 1) getVariable [QGVAR(jammedWeapons), []])}
-    && {!(currentMuzzle (_this select 1) in ((_this select 1) getVariable [QEGVAR(safemode,safedWeapons), []]))}
-  };
 
-  _statement = {
-    playSound3D ["a3\sounds_f\weapons\Other\dry9.wss", _this select 0];
-
-    if (!(missionNamespace getVariable [QGVAR(knowAboutJam), false]) && {(_this select 1) ammo currentWeapon (_this select 1) > 0} && {GVAR(DisplayTextOnJam)}) then {
-      [localize LSTRING(WeaponJammed)] call EFUNC(common,displayTextStructured);
-      GVAR(knowAboutJam) = true;
+    private _condition = {
+        [_this select 1] call EFUNC(common,canUseWeapon)
+        && {currentMuzzle (_this select 1) in ((_this select 1) getVariable [QGVAR(jammedWeapons), []])}
+        && {!(currentMuzzle (_this select 1) in ((_this select 1) getVariable [QEGVAR(safemode,safedWeapons), []]))}
     };
-  };
 
-  _condition2 = {
-    currentWeapon (_this select 1) in ((_this select 1) getVariable [QGVAR(jammedWeapons), []])
-  };
+    private _statement = {
+        playSound3D ["a3\sounds_f\weapons\Other\dry9.wss", _this select 0];
```

Comment:
Those are not for randomization, but specific weapons use a specific sound sample. We should probably read the sample from config (possibly with caching)